### PR TITLE
Make UniformSampling inherit from FilterIndices instead of Filter

### PR DESCRIPTION
**Pull request by QiuYilin**
finish https://github.com/PointCloudLibrary/pcl/issues/6051

Please let me know if there is anything that needs to be modified

### CODE REVIEW

File: filters/include/pcl/filters/impl/uniform_sampling.hpp [hunk 1]
```diff
@@ -126,26 +119,30 @@ pcl::UniformSampling<PointT>::applyFilter (PointCloud &output)
...
     {
       if (extract_removed_indices_)
-        removed_indices_->push_back ((*indices_)[cp]);
+        (*removed_indices_)[rii++] = (*indices_)[cp];
     }
   }
 
   // Second pass: go over all leaves and copy data
-  output.resize (leaves_.size ());
   std::size_t cp = 0;
```
Comment (larshg) [inline filters/include/pcl/filters/impl/uniform_sampling.hpp:133]:
```suggestion
```

Comment (larshg) [reply filters/include/pcl/filters/impl/uniform_sampling.hpp:133]: I think this one is not used.

File: filters/include/pcl/filters/impl/uniform_sampling.hpp [hunk 2]
```diff
@@ -43,19 +43,13 @@
 
 //////////////////////////////////////////////////////////////////////////////////////////////////////////////////
 template <typename PointT> void
-pcl::UniformSampling<PointT>::applyFilter (PointCloud &output)
+pcl::UniformSampling<PointT>::applyFilter (Indices &indices)
```
Comment (larshg) [inline filters/include/pcl/filters/impl/uniform_sampling.hpp:46]: I think we need to keep the old method still but add deprecation warning for removal in 1.17.

Comment (QiuYilin) [reply filters/include/pcl/filters/impl/uniform_sampling.hpp:46]: This function is called through pcl::FilterIndices<PointT>::applyFilter (PointCloud &output), so it should not conflict with users' old code.

Comment (mvieth) [reply filters/include/pcl/filters/impl/uniform_sampling.hpp:46]: `FilterIndices` has [that method](https://pointclouds.org/documentation/classpcl_1_1_filter_indices.html#a237872d14a14ba3881592d3106cf4338) which should take its place, right?

File: filters/include/pcl/filters/impl/uniform_sampling.hpp (hunk 1 above)
Comment (QiuYilin) [reply filters/include/pcl/filters/impl/uniform_sampling.hpp]: updated

File: test/filters/test_uniform_sampling.cpp [hunk 3]
```diff
@@ -65,13 +65,71 @@ TEST(UniformSampling, extractRemovedIndices)
...
   pcl::PointCloud<pcl::PointXYZ> output;
+
+  //cloud
   us.filter(output);
 
+  //indices
+  pcl::Indices indices;
+  us.filter(indices);
+
+  for (std::size_t i = 0; i < indices.size(); ++i)
```
[the remark this answers is not in the document]
Comment (QiuYilin) [reply test/filters/test_uniform_sampling.cpp]: updated

File: filters/include/pcl/filters/impl/uniform_sampling.hpp (hunk 2 above)
Comment (larshg) [reply filters/include/pcl/filters/impl/uniform_sampling.hpp:46]: Ahh, yeah ofc.

File: filters/include/pcl/filters/uniform_sampling.h [hunk 4]
```diff
@@ -69,14 +71,13 @@ namespace pcl
     using Filter<PointT>::getClassName;
 
     public:
-      using Ptr = shared_ptr<UniformSampling<PointT> >;
```
Comment (mvieth) [inline filters/include/pcl/filters/uniform_sampling.h:72]: Why did you remove this?

File: filters/include/pcl/filters/impl/uniform_sampling.hpp [hunk 5]
```diff
@@ -43,19 +43,13 @@
...
-  // Has the input dataset been set already?
-  if (!input_)
-  {
-    PCL_WARN ("[pcl::%s::applyFilter] No input dataset given!\n", getClassName ().c_str ());
-    output.width = output.height = 0;
-    output.clear ();
-    return;
-  }
+  // The arrays to be used
+  indices.resize (indices_->size ());
```
Comment (mvieth) [inline filters/include/pcl/filters/impl/uniform_sampling.hpp:49]: I think `indices` could be instead resized to `leaves_.size()` just before the second pass. That way, we have a better, "tighter" estimate of the final size.

File: filters/include/pcl/filters/impl/uniform_sampling.hpp [hunk 6]
```diff
@@ -43,19 +43,13 @@
 
 //////////////////////////////////////////////////////////////////////////////////////////////////////////////////
 template <typename PointT> void
-pcl::UniformSampling<PointT>::applyFilter (PointCloud &output)
+pcl::UniformSampling<PointT>::applyFilter (Indices &indices)
 {
-  // Has the input dataset been set already?
-  if (!input_)
```
Comment (mvieth) [inline filters/include/pcl/filters/impl/uniform_sampling.hpp:49]: Might make sense to keep this check? And clear `indices` and `removed_indices_` if input has not been set?

Comment (QiuYilin) [reply filters/include/pcl/filters/impl/uniform_sampling.hpp:49]: https://github.com/PointCloudLibrary/pcl/blob/master/common/include/pcl/impl/pcl_base.hpp#L143 There is already a check in the compute initialization process, and other filters don't have this part.

File: filters/include/pcl/filters/impl/uniform_sampling.hpp (hunk 5 above)
Comment (QiuYilin) [reply filters/include/pcl/filters/impl/uniform_sampling.hpp:49]: leaves_ is a std::unordered_map without init size. This also results in the order of the indexes calculated each time being uncertain.

File: filters/include/pcl/filters/uniform_sampling.h (hunk 4 above)
Comment (QiuYilin) [reply filters/include/pcl/filters/uniform_sampling.h]: It is a mistake. Updated.

File: filters/include/pcl/filters/impl/uniform_sampling.hpp (hunk 5 above)
Comment (mvieth) [reply filters/include/pcl/filters/impl/uniform_sampling.hpp:49]: Sorry, I might have been unclear. I meant in line 132, just before iterating over `leaves_` in the for-loop. As far as I can see, there is nothing inserted into `indices` before that for-loop.
The order of the indices is random, not matter where `indices` is resized, right?

File: filters/include/pcl/filters/impl/uniform_sampling.hpp (hunk 6 above)
Comment (mvieth) [reply filters/include/pcl/filters/impl/uniform_sampling.hpp:49]: You are right, thanks!

File: filters/include/pcl/filters/impl/uniform_sampling.hpp (hunk 5 above)
Comment (QiuYilin) [reply filters/include/pcl/filters/impl/uniform_sampling.hpp:49]: Updated

**Review by mvieth — DISMISSED** [review]
Looks good to me, thank you!

File: filters/include/pcl/filters/impl/uniform_sampling.hpp [hunk 7]
```diff
@@ -79,27 +73,26 @@ pcl::UniformSampling<PointT>::applyFilter (PointCloud &output)
...
-  for (std::size_t cp = 0; cp < indices_->size (); ++cp)
+  for (const auto& cp : *indices_)
   {
     if (!input_->is_dense)
     {
       // Check if the point is invalid
-      if (!std::isfinite ((*input_)[(*indices_)[cp]].x) || 
-          !std::isfinite ((*input_)[(*indices_)[cp]].y) || 
-          !std::isfinite ((*input_)[(*indices_)[cp]].z))
+      if (!std::isfinite ((*input_)[cp].x) ||
```
Comment (larshg) [inline filters/include/pcl/filters/impl/uniform_sampling.hpp:82]: Could use [pcl::isXYZFinite](https://pointclouds.org/documentation/namespacepcl.html#af0863d8e9957d47a89538a631122efd9).

Comment (QiuYilin) [reply filters/include/pcl/filters/impl/uniform_sampling.hpp]: updated

**Review by larshg — APPROVED** [review]
